docs/coding-guidelines/interop-guidelines.md

```diff
@@ -204,3 +188,273 @@ Guidelines for shim C++ API:
+
+[1]: https://msdn.microsoft.com/en-us/library/system.runtime.interopservices.dllimportattribute.aspx
+[2]: https://msdn.microsoft.com/en-us/library/system.runtime.interopservices.dllimportattribute.preservesig.aspx
+[3]: https://msdn.microsoft.com/en-us/library/system.runtime.interopservices.dllimportattribute.setlasterror.aspx
+[4]: https://msdn.microsoft.com/en-us/library/system.runtime.interopservices.dllimportattribute.exactspelling.aspx
+[5]: https://msdn.microsoft.com/en-us/library/system.runtime.interopservices.dllimportattribute.charset.aspx
+
+Strings
+-------
+
+When the CharSet is Unicode or the argument is explicitly marked as `[MarshalAs(UnmanagedType.LPWSTR)]` _and_ the string is passed by value (not `ref` or `out`) the string will be pinned and used directly by native code (rather than copied).
+
+Remember to mark the `[DllImport]` as `Charset.Unicode` unless you explicitly want ANSI treatment of your strings.
+
+Be sure to allow space for null terminator if necessary.
+
+**[AVOID]** `StringBuilder` marshalling. It *always* creates a native buffer copy (see `ILWSTRBufferMarshaler`). As such it can be extremely inefficient. Take the typical
+scenario of calling a Windows API that takes a string:
+
+1. Create a SB of the desired capacity (allocates managed capacity) **{1}**
+2. Invoke
+   1. Allocates a native buffer **{2}**  
+   2. Copies the contents if `[In]` _(the default for a `StringBuilder` parameter)_  
+   3. Copies the native buffer into a newly allocated managed array if `[Out]` **{3}** _(also the default for `StringBuilder`)_  
+3. `ToString()` allocates yet another managed array **{4}**
+
+That is *{4}* allocations to get a string out of native code. The best you can do to limit this is to reuse the `StringBuilder`
+in another call but this still only saves *1* allocation. It is much better to use and cache a character buffer from `ArrayPool`- you can then get down to just the allocation for the `ToString()` on subsequent calls. 
+
+The other issue with `StringBuilder` is that it always copies the return buffer back up to the first null. If the passed back string isn't terminated or is a double-null-terminated string your P/Invoke is incorrect at best.
+
+If you *do* use `StringBuilder` one last gotcha is that the capacity does **not** include a hidden null which is always accounted for in interop. It is pretty common for people to get this wrong as most APIs want the size of the buffer *including* the null. This can result in wasted/unnecessary allocations.
+
+**[USE]** Char arrays from `ArrayPool` or `StringBuffer` or `stackalloc`
```

Comment:
I added stackalloc here. Probably, we should explicitly talk about buffer allocation in general (not just char buffers)